### PR TITLE
Clean up extension validation

### PR DESCRIPTION
**Pull request by slavapestov**
I started looking at a fix for rdar://44429394 and ended up doing this first...

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/AST/DeclContext.cpp [hunk 1]
```diff
@@ -92,7 +92,11 @@ GenericTypeParamType *DeclContext::getProtocolSelfType() const {
     const_cast<ProtocolDecl *>(proto)->createGenericParamsIfMissing();
   }
 
-  return getGenericParamsOfContext()->getParams().front()
+  auto *genericParams = getGenericParamsOfContext();
+  if (genericParams == nullptr)
+    return nullptr;
```
Comment (jrose-apple) [inline lib/AST/DeclContext.cpp]: When does this come up for a protocol extension?

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -4689,7 +4689,25 @@ void TypeChecker::finalizeDecl(ValueDecl *decl) {
...
+///
+/// whereas \c GX2 and \c GX3 are not pass-through because \c GX2 has
+/// different type parameters and \c GX3 doesn't pass its type parameters
+/// directly through.
+///
+/// \code
+/// typealias GX2<A> = X<A, A>
+/// typealias GX3<A, B> = X<B, A>
+/// \endcode
+bool isPassThroughTypealias(TypeAliasDecl *typealias) {
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: `static`?

File: lib/AST/DeclContext.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/AST/DeclContext.cpp]: Invalid code, like when an extension is inside of another decl, we don't build the generic parameter list for it anymore.

File: lib/Sema/TypeCheckDecl.cpp (hunk 2 above)
Comment (slavapestov) [reply lib/Sema/TypeCheckDecl.cpp]: Right...

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/AST/DeclContext.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/AST/DeclContext.cpp]: But we still bind the underlying type? Okay.

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -4877,31 +4895,78 @@ void TypeChecker::validateExtension(ExtensionDecl *ext) {
...
   auto extendedType = ext->getExtendedType();
 
-  if (extendedType.isNull() || extendedType->hasError())
+  // Hack to allow extending a generic typealias.
+  if (auto *unboundGeneric = extendedType->getAs<UnboundGenericType>()) {
+    if (auto *aliasDecl = dyn_cast<TypeAliasDecl>(unboundGeneric->getDecl())) {
+      auto extendedNominal = aliasDecl->getDeclaredInterfaceType()->getAnyNominal();
+      if (extendedNominal) {
+        extendedType = extendedNominal->getDeclaredType();
+        if (!isPassThroughTypealias(aliasDecl))
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: I suppose this is where we should diagnose non-passthrough `typealias` declarations that will turn into extensions of the underlying nominal type (and not become parameterized extensions magically).

**Review by DougGregor — APPROVED** [review]
Thanks!

**Pull request merged**